### DISCUSSION TITLE
Support other explorer to verify smart contract with verify plugin

### DISCUSSION THREAD

**Discussion by huihzhao** [NomicFoundation/hardhat#4212] [Ideas]
I am from opbnbscan explorer team and I`m doing the compatibility testing for our explorer API of hardhat. but seems the verify plugin only works with etherscan and blockscout. Just wondering how to support other explorers like ours.

**Comment by schaable** [comment]
Hi @huihzhao! It depends on your explorer API. If it has the same interface than the Etherscan API, you can add it as a custom chain like this:

```
// hardhat.config.ts

const config: HardhatUserConfig = {
  etherscan: {
    apiKey: {
      <explorer-name>: "<explorer-api-key>"
    },
    customChains: [
      {
        network: <explorer-name>,
        chainId: <explorer-chain-id>,
        urls: {
          apiURL: <explorer-api-url>,
          browserURL: <explorer-browser-url>
        }
      }
    ]
  }
}
```

However, if the interface is different, you would need to extend the plugin with your own subtasks. You can package it in your own plugin, for example `hardhat-verify-bnb`. Let me know if you want to go that route and we can discuss the details.

**Reply by huihzhao** [reply]
Thanks for the clarification, it seems our API is not fully compatible with the ethscan ones. That is to say we may want to add the subtask. May I know how to?

**Reply by huihzhao** [reply]
It is better that you can have a doc for this. I believe more chains will have similar requirements. 😄

**Reply by schaable** [reply]
To add a new verification provider, you would need to override the `verify:get-verification-subtasks` subtask and add your subtask to the pipeline. You can find the definition [here](https://github.com/NomicFoundation/hardhat/blob/dd2297db30fba999917c247b8e3612c9d8315e76/packages/hardhat-verify/src/index.ts#L213). Please note that there's an [open PR](https://github.com/NomicFoundation/hardhat/pull/4199) that may change the code structure slightly, although it should not affect this process.

For example:

```
subtask("verify:get-verification-subtasks", async (taskArgs, hre, runSuper) => {
  const verificationSubtasks = await runSuper(taskArgs);

  verificationSubtasks.push("verify:opbnbscan");

  return verificationSubtasks;
});
```

Then, you would need to add the `verify:opbnbscan` subtask with your custom validation logic.

**Reply by schaable** [reply]
> It is better that you can have a doc for this. I believe more chains will have similar requirements. 😄

I don't think it's necessary, considering there's already a [document explaining how to override tasks](https://hardhat.org/hardhat-runner/docs/advanced/create-task#overriding-tasks). I agree that there should be a code example there, and I will work on [adding it](https://github.com/NomicFoundation/hardhat/issues/4215). If we encounter similar questions in the future, we can consider providing a tailored explanation.
